### PR TITLE
Document Roslyn Diagnostics Security Analyzer

### LINKED ISSUE #1669: Document Roslyn Diagnostics Security Analyzers
Real mail from a customer today:

> One of our security requirements is to run static analysis on our product’s code. The code analysis tool in VS 2015 can’t analyze .net core projects. What would be the best tool for this? We are currently using .net core 1.1

Our docs need to answer this question. This is our guidance: https://blogs.msdn.microsoft.com/secdevblog/2016/03/30/roslyn-diagnostics-security-analyzers-overview/

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes #1669 

Note that the configuration instructions in the blog post referenced in issue #1669 are not longer needed.

This PR is dependent on #3908

### CODE REVIEW

File: docs/standard/security-analyzer.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ms.topic: article
+ms.prod: .net
+ms.technology: dotnet-standard
+ms.devlang: dotnet
+---
+
+# The .NET Security Analyzers
+
+You can use the .NET Security Analyzers to find potential security risks
+in your code. These analyzers will find potential issues, and suggest
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: no comma after "issues"

File: docs/standard/security-analyzer.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+fixes to those issues.
+
+The analyzers run interactively in Visual Studio, as you write your code,
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: no comma either before as and after code.

File: docs/standard/security-analyzer.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+or as part of a CI build. You'll want to add the analyzer to your project as
+early as possible in your development. The sooner you find any potential issues
+in your code, the easier it will be to fix. However, you can add it at any time
+in the development cycle. It will find any issues with the existing code, and
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: no comma before and

File: docs/standard/security-analyzer.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+will start warning about new issues as you keep developing.
+
+## Installing and configuring the .NET Security Analyzer
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: Analyzer --> Analyzers

File: docs/standard/security-analyzer.md [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+The .NET Security Analyzers must be installed as a NuGet package on every
+project where you want them to run. Only one developer needs to add them
+to the project. The analyzer package is a project dependency, and will run
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: no comma after dependency

File: docs/standard/security-analyzer.md [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+on every developer's machine once they have the updated solution.
+
+The .NET Security Analyzers are part of the [Microsoft.DesktopAnalyzers](https://www.nuget.org/packages/Desktop.Analyzers/)
+NuGet package. This package provides a smaller set of analyzers
+that includes the security analyzers. In most cases, you'll want
+the [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) NuGet package.
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: Are the SecurityAnalyzers also included in FxCopAnalyzers? In any case, the relationship is unclear.

File: docs/standard/security-analyzer.md [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+the [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) NuGet package. 
+The FxCopAnalyzers package contains many more analyzers that
+help your team enforce the rules recommended by FxCop.
+
+To install it, right-click on the project, and select "Manage Dependencies".
+From the NuGet explorer, search for "Desktop Analyzers", or if you prefer,
+"Fx Cop Analyzers". Install the latest stable version in all projects in your
+solution.
+
+Once the NuGet package is installed run a build, and you'll see a list of
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: comma after installed
no comma after build

File: docs/standard/security-analyzer.md [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+warnings for the three security related issues found by these analyzers:
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: what is the analyzer running against to produce these warnings? Does it come with a project that produces the warnings?

File: docs/standard/security-analyzer.md [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+
+- [CA2153](~/visualstudio/code-quality/ca2153-avoid-handling-corrupted-state-exceptions): Avoid handling corrupted state exceptions
+- [CA5350](~/visualstudio/code-quality/ca5350-do-not-use-weak-cryptographic-algorithms): Do not use weak cryptographic algorithms
+- [CA5351](~/visualstudio/code-quality/ca5351-do-not-use-broken-cryptographic-algorithms): Do not use broken cryptographic algorithms
+
+Code that triggers any of these analyzers generates a warning by default.
+
+## Roslyn based analyzers vs. Code Analysis libraries
```
Comment (rpetrusha) [inline docs/standard/security-analyzer.md]: Here and throughout, Roslyn based --> Roslyn-based

**Review by rpetrusha — CHANGES_REQUESTED** [review]
Left some questions and comments, @BillWagner.

File: docs/standard/security-analyzer.md (hunk 8 above)
Comment (BillWagner) [reply docs/standard/security-analyzer.md]: I re-wrote it to be more clear.

File: docs/standard/security-analyzer.md (hunk 6 above)
Comment (BillWagner) [reply docs/standard/security-analyzer.md]: I re-wrote it to be more clear.

**Comment by BillWagner** [comment]
I updated this based on your feedback @rpetrusha

**Comment by BillWagner** [comment]
Do you want to provide a tech review @mavasani

File: docs/standard/security-analyzer.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+will start warning about new issues as you keep developing.
+
+## Installing and configuring the .NET Security Analyzers
+
+The .NET Security Analyzers must be installed as a NuGet package on every
+project where you want them to run. Only one developer needs to add them
+to the project. The analyzer package is a project dependency and will run
+on every developer's machine once they have the updated solution.
+
+The .NET Security Analyzers are delivered in the [Microsoft.DesktopAnalyzers](https://www.nuget.org/packages/Desktop.Analyzers/)
```
Comment (mavasani) [inline docs/standard/security-analyzer.md]: Desktop analyzers package is no longer maintained, and we should not recommend anyone to install them. We have refactored the analyzers into following packages:

**Core analyzer packages**
1. [Microsoft.CodeQuality.Analyzers](https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers): Common code quality rules, which are not specific to any .net framework or APIs. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.Analyzers.md)
2. [Microsoft.NetCore.Analyzers](https://www.nuget.org/packages/Microsoft.NetCore.Analyzers): Analyzers specific to netcore APIs. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetCore.Analyzers/Microsoft.NetCore.Analyzers.md)
3. [Microsoft.NetFramework.Analyzers](https://www.nuget.org/packages/Microsoft.NetFramework.Analyzers): Analyzers specific to netfx APIs. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.NetFramework.Analyzers/Microsoft.NetFramework.Analyzers.md)
4. [Text.Analyzers](https://www.nuget.org/packages/Text.Analyzers): Analyzers specific to text APIs, e.g. for analyzer and code fix authors. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Text.Analyzers/Text.Analyzers.md)
5. [Microsoft.CodeAnalysis.Analyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.Analyzers): Analyzers specific to usage of Roslyn/Microsoft.CodeAnalysis APIs, e.g. for analyzer and code fix authors. See documentation for rules included in this package [here](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.md)

**Aggregate analyzer packages**
1. [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers/): Analyzers aimed at replacement of legacy FXCop rules. This package is an aggregate of the following analyzer packages:
   1. Microsoft.CodeQuality.Analyzers
   2. Microsoft.NetCore.Analyzers
   3. Microsoft.NetFramework.Analyzers
   4. Text.Analyzers

File: docs/standard/security-analyzer.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+help your team enforce the rules recommended by FxCop.
+
+To install it, right-click on the project, and select "Manage Dependencies".
+From the NuGet explorer, search for "Desktop Analyzers", or if you prefer,
+"Fx Cop Analyzers". Install the latest stable version in all projects in your
+solution.
+
+Once the NuGet package is installed, build your solution. The analyzers examine
+the code in your solution and provide you with a list of
+warnings for the three security related issues:
```
Comment (mavasani) [inline docs/standard/security-analyzer.md]: > three security related issues

These rules have been moved to `Microsoft.NetFramework.Analyzers`

File: docs/standard/security-analyzer.md [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+- [CA5351](~/visualstudio/code-quality/ca5351-do-not-use-broken-cryptographic-algorithms): Do not use broken cryptographic algorithms
+
+Code that triggers any of these analyzers generates a warning by default.
+
+## Roslyn-based analyzers vs. Code Analysis libraries
+
+The user experience for Roslyn-based analyzers is different than that of the
+Code Analysis libraries like the older versions of FxCop and the Security analysis
+tools.  You don't need to explicitly run the Roslyn-based analyzers. There's no need to use the "Run Code Analysis"
+menu items on the "Analyze" menu in Visual Studio. Roslyn-based analyzers run asychronously as you work.
```
Comment (mavasani) [inline docs/standard/security-analyzer.md]: You probably want to also mention that doing a "Build" in Visual Studio or command line will also cause the analyzers to run and report diagnostics, just like compiler diagnostics. In fact, analyzers are essentially plugins that C# and VB compilers invoke during build and live analysis, and then it aggregates and outputs the compiler and analyzer diagnostics.

**Comment by BillWagner** [comment]
Adding @richlander (wrote #1669) and @OliaG (wrote #3908)

Reading @mavasani 's comments, I think the focus here is wrong based on the direction and the packaging. The analyzers that were part of the security set are now in the NetFramework and NetCore packages, where applicable.  That means this article would need an update to direct users to different packages for Core and Framework projects.

I'll propose that a better way to fix #1669 is to write an article that discusses using the first 4 analyzers @mavasani mentions in his comment above, and the aggregate FxCop analzyer package.

I could rework what's here (including changing the file name) and have it done in the next sprint.

Upvote if you agree, downvote with suggestions if you disagree.

**Comment by BillWagner** [comment]
Adding the [WIP] label because some larger changes may be coming.

File: docs/standard/security-analyzer.md [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+The .NET Security Analyzers are delivered in the [Microsoft.NetFramework.Analyzers](https://www.nuget.org/packages/Microsoft.NetFramework.Analyzers/)
+NuGet package. This package provides only the analyzers specific to the .NET Framework, which
+includes the security analyzers. In most cases, you'll want
+the [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) NuGet package. 
+The FxCopAnalyzers aggregate package contains all the security analyzers included in the
+Desktop.Analyzers package as well as other analyzers that
+help your team enforce the rules recommended by FxCop.
+
+To install it, right-click on the project, and select "Manage Dependencies".
```
Comment (richlander) [inline docs/standard/security-analyzer.md]: Is there a separate topic you can point to that walks through the details of NuGet package installation more generally?

File: docs/standard/security-analyzer.md [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+From the NuGet explorer, search for "NetFramework Analyzers", or if you prefer,
+"Fx Cop Analyzers". Install the latest stable version in all projects in your
+solution.
+
+Once the NuGet package is installed, build your solution. The analyzers examine
```
Comment (richlander) [inline docs/standard/security-analyzer.md]: Maybe this is a start to a "Using the .NET Security Analyzers" section?

File: docs/standard/security-analyzer.md [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+the code in your solution and provide you with a list of
+warnings for the three security related issues:
+
+- [CA2153](~/visualstudio/code-quality/ca2153-avoid-handling-corrupted-state-exceptions): Avoid handling corrupted state exceptions
+- [CA5350](~/visualstudio/code-quality/ca5350-do-not-use-weak-cryptographic-algorithms): Do not use weak cryptographic algorithms
+- [CA5351](~/visualstudio/code-quality/ca5351-do-not-use-broken-cryptographic-algorithms): Do not use broken cryptographic algorithms
+
+Code that triggers any of these analyzers generates a warning by default.
```
Comment (richlander) [inline docs/standard/security-analyzer.md]: Seems like a good opportunity to a screenshot

File: docs/standard/security-analyzer.md [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+## Roslyn-based analyzers vs. Code Analysis libraries
```
Comment (richlander) [inline docs/standard/security-analyzer.md]: This section threw me off a bit.  I didn't realize at first that this was a comparison and not more info about the main topic.  I don't have specific suggestions on how to solve that.

**Comment by richlander** [comment]
The other option is to describe the general capability (which is close to what you have here) and then link to platform-specific docs for the specific instructions for using the analyzers. The benefit of this approach would be two-folder: separation of concerns and you'd get readers in both the standard and non-standard docs. Something to consider.

**Comment by mairaw** [comment]
Haven’t reviewed this yet, but you might wanna put this topic in the analyzers folder we’re creating in the dependent PR. We also need to create an index page either here or there or on a different PR.

File: docs/standard/security-analyzer.md (hunk 12 above)
Comment (BillWagner) [reply docs/standard/security-analyzer.md]: Moved to the index.md file in the analyzers directory as this applies to all Roslyn-based analyzers.

File: docs/standard/security-analyzer.md (hunk 10 above)
Comment (BillWagner) [reply docs/standard/security-analyzer.md]: I'll expand this PR to cover NetFramework.Analyzers.  I've opened the following new issues for the others that are part of the FxCopAnalyzers package:

#4214
#4215
#4216

Note that the CodeAnalysis package will be covered as part of the Roslyn API documentation.

File: docs/standard/security-analyzer.md (hunk 13 above)
Comment (BillWagner) [reply docs/standard/security-analyzer.md]: I added this to a new index.md file in the /standard/analyzers folder. That will be applicable to the changes in #3908

File: docs/standard/security-analyzer.md (hunk 16 above)
Comment (BillWagner) [reply docs/standard/security-analyzer.md]: I moved it to a note in the index.md file that I added. I hope that sets it as a "sidebar" to the main topic.

**Comment by BillWagner** [comment]
I updated and responded to all the feedback. This is ready for another review.

**Review by rpetrusha — CHANGES_REQUESTED** [review]
This topic looks really good, @BillWagner, although I've left a number of comments.

File: docs/standard/analyzers/framework-analyzer.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+
+![issues reported by the framework analyzer](./media/framework-analyzers-2.png)
+
+As you write code, you'll see squiggles underneath any potential issue in your code.
+Hover over any issue, and you'll be shown details about the issue, and suggestions
+for any possible fix, as shown in the following image:
+
+![interactive report of issues found by the framework analyzer](./media/framework-analyzers-1.png)
+
+## images to add:
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/standard/analyzers/framework-analyzer.md]: oops.  :blush:

**Comment by BillWagner** [comment]
Thanks @rpetrusha I've responded to all the comments.

File: docs/standard/analyzers/framework-analyzer.md [hunk 18]
```diff
@@ -0,0 +1,158 @@
...
+
+The .NET Framework Analyzer is delivered in the [Microsoft.NetFramework.Analyzers](https://www.nuget.org/packages/Microsoft.NetFramework.Analyzers/)
+NuGet package. This package provides only the analyzers specific to the .NET Framework, which
+includes security analyzers. In most cases, you'll want
+the [Microsoft.CodeAnalysis.FxCopAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.FxCopAnalyzers) NuGet package. 
+The FxCopAnalyzers aggregate package contains all the framework analyzers included in the
+Framework.Analyzers package as well as the following analyzers:
+- [Microsoft.CodeQuality.Analyzers](https://www.nuget.org/packages/Microsoft.CodeQuality.Analyzers): Provides general guidance and guidance for .NET Standard APIs
+- [Microsoft.NetCore.Analyzers](https://www.nuget.org/packages/Microsoft.NetCore.Analyzers): Provides analyzers specific to .NET Core APIs.
+- [Text.Analyzers](https://www.nuget.org/packages/Text.Analyzers): Provides guidance for text included as code including comments.
```
Comment (rpetrusha) [inline docs/standard/analyzers/framework-analyzer.md]: nit: there should be a comma after "code"

**Review by rpetrusha — APPROVED** [review]
Great job on the topic and the revision, @BillWagner. I've left one nit, which you can choose to address or ignore.

**Pull request merged**